OpenRA.Mods.Common/Traits/World/MovementClass.cs

```diff
@@ -0,0 +1,341 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2018 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Mods.Common.Pathfinder;
+using OpenRA.Primitives;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Flags]
+	public enum CellConditions
+	{
+		None = 0,
+		TransientActors,
+		BlockedByMovers,
+		All = TransientActors | BlockedByMovers
+	}
+
+	public static class CellConditionsExts
+	{
+		public static bool HasCellCondition(this CellConditions c, CellConditions cellCondition)
+		{
+			// PERF: Enum.HasFlag is slower and requires allocations.
+			return (c & cellCondition) == cellCondition;
+		}
+	}
+
+	public static class CustomMovementLayerType
+	{
+		public const byte Tunnel = 1;
+		public const byte Subterranean = 2;
+		public const byte Jumpjet = 3;
+		public const byte ElevatedBridge = 4;
+	}
+
+	[Desc("Used by Mobile. Attach these to the world actor. You can have multiple variants by adding @suffixes.")]
+	public class MovementClassInfo : ITraitInfo
+	{
+		[Desc("Class ID.")]
+		public readonly string Class = "default";
+
+		public readonly int WaitAverage = 5;
+
+		public readonly int WaitSpread = 2;
+
+		[Desc("Allow multiple (infantry) units in one cell.")]
+		public readonly bool SharesCell = false;
+
+		[Desc("Can the actor be ordered to move in to shroud?")]
+		public readonly bool MoveIntoShroud = true;
+
+		[Desc("Can this unit move underground?")]
+		public readonly bool Subterranean = false;
```

Comment:
Can you please define subclasses for `SubterraneanMovementClassInfo` and the other custom types?

The checks for `if (mc.Subterranean)` can then be replaced by `if (mc is SubterraneanMovementClassInfo)` or
```
var smc = mc as SubterraneanMovementClassInfo;
if (smc != null)
```
if the Subterranean specific info is needed.